### PR TITLE
project-watcher can request eventual callback, or immediately call ca…

### PR DESCRIPTION
**Pull request by benjiwheeler**
Addresses a bug introduced in https://github.com/LLK/scratch-gui/pull/3638 where user cannot click on See Community button to transition from editor to project page, if user does now own project.

Revises `ProjectWatcher` component so that it can be set to callback a function after waiting for project to finish saving, *or* to immediately call the callback.

Also ensures that saves before transition are auto-saves, not manual saves.

### CODE REVIEW

File: src/components/menu-bar/menu-bar.jsx [hunk 1]
```diff
@@ -681,6 +688,7 @@ MenuBar.propTypes = {
     authorId: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
     authorThumbnailUrl: PropTypes.string,
     authorUsername: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
+    autoUpdateProject: PropTypes.func,
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/menu-bar/menu-bar.jsx]: Yes, it should.

File: src/components/menu-bar/menu-bar.jsx [hunk 2]
```diff
@@ -749,6 +757,7 @@ const mapStateToProps = state => {
 };
 
 const mapDispatchToProps = dispatch => ({
+    autoSave: () => dispatch(autoUpdateProject()),
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/menu-bar/menu-bar.jsx]: there are "auto" saves and "manual" saves, distinguished by project-state and project-saver-hoc. Manual saves are deliberately clicked by the user; auto saves are not. 

The distinction matters for alerts.

File: src/containers/project-watcher.jsx [hunk 3]
```diff
@@ -20,18 +20,24 @@ class ProjectWatcher extends React.Component {
...
-    setRequesting () {
-        this.setState({
-            requesting: true
+    fulfillRequest () {
+        this.props.onShowingWithId();
+        this.setState({ // eslint-disable-line react/no-did-update-set-state
+            requesting: false
         });
     }
+    setRequesting (waitForRequest) {
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/containers/project-watcher.jsx]: Done. Renamed a bunch of stuff with @chrisgarrity and added JSDoc.

File: src/containers/project-watcher.jsx [hunk 4]
```diff
@@ -7,46 +7,60 @@ import {
...
+ * ProjectWatcher passes a waitForUpdate function to its children, which they can call
+ * to set ProjectWatcher to request that it call the onDoneUpdating callback when
+ * project is no longer updating.
+ */
 class ProjectWatcher extends React.Component {
     constructor (props) {
         super(props);
         bindAll(this, [
-            'setRequesting'
+            'waitForSaving'
```
Comment (chrisgarrity) [inline src/containers/project-watcher.jsx]: typo? `waitForUpdate`

File: src/containers/project-watcher.jsx [hunk 5]
```diff
@@ -7,46 +7,60 @@ import {
...
-            });
+        if (this.state.waiting && this.props.isShowingWithId && !prevProps.isShowingWithId) {
+            this.fulfillRequest();
         }
-
     }
-    setRequesting () {
-        this.setState({
-            requesting: true
+    fulfillRequest () {
```
Comment (chrisgarrity) [inline src/containers/project-watcher.jsx]: Another renaming suggestion (you can ignore). Request is just so overloaded. Would fulfillAction make sense instead as it's fulfilling the user's action to share or see community.

File: src/containers/project-watcher.jsx (hunk 4 above)
Comment (benjiwheeler) [reply src/containers/project-watcher.jsx]: Fixed

File: src/containers/project-watcher.jsx (hunk 5 above)
Comment (benjiwheeler) [reply src/containers/project-watcher.jsx]: How about "fulfill()"?

**Pull request merged**